R/derive_interp_records.R

```diff
@@ -59,16 +63,33 @@ derive_interp_records <- function(dataset, parameter = "WEIGHT") {
   }
 
   stopifnot("AGE" %in% colnames(dataset))
-  stopifnot("SEX" %in% colnames(dataset))
+  stopifnot(expr(!!by_vars) %in% colnames(dataset))
   stopifnot("L" %in% colnames(dataset))
   stopifnot("M" %in% colnames(dataset))
   stopifnot("S" %in% colnames(dataset))
   if (parameter == "BMI") {
     stopifnot("P95" %in% colnames(dataset))
     stopifnot("Sigma" %in% colnames(dataset))
   }
-  arrange(dataset, SEX, AGE)
-  fapp <- function(v) approx(dataset$AGE, v, xout = 730:7305)$y
-  x <- lapply(dataset[, metadata_vars], fapp)
-  as.data.frame(do.call(bind_cols, x)) %>% filter(!is.na(AGE))
+
+  arrange(dataset, !!!by_vars, AGE)
+
+  fapp <- function(v, age) {
+    approx(age, v, xout = seq(min(age), max(age)))$y
+  }
+
+  # Apply the function within each group and combine the results
+  result <- dataset %>%
+    group_by(!!!by_vars) %>%
+    do({
+      age <- .$AGE
+      x <- lapply(.[, metadata_vars], fapp, age = age)
+      as.data.frame(do.call(bind_cols, x))
+    }) %>%
+    ungroup() %>%
+    # Keep patients >= 2 yrs till 20 yrs - Remove duplicates for 730 Days old which
+    # must come from WHO metadata only
+    filter(!is.na(AGE) & AGE >= 730.5 & AGE <= 7305)
```

Comment:
The step for ` # Keep patients >= 2 yrs till 20 yrs - Remove duplicates for 730 Days old...` should not be in this function. Better to just leave as `filter(!is.na(AGE))`, because what about the users who don't want to follow our rules here around only using CDC growth charts for >=2 years. Our functions need to be flexible to be applied to whatever choice of metadata the user likes, e.g. maybe they only want to use interpolated CDC for all age ranges (and not WHO at all). Then in our template only we can make such recommended choices.

Reply:
Moved back to the ADVS template